lls/cache.c

```diff
@@ -0,0 +1,481 @@
+		 * it's unlikely but possible for multiple for quick
+		 * timeouts and updates to cause multiple callbacks
+		 * before the requesting block is able to remove its
+		 * LLS_REQ_RESOLVE request.
+		 */
+		else if (!record->holds[i].notified && activated) {
+			record->holds[i].cb(map, record->holds[i].arg,
+				LLS_REPLY_RESOLUTION);
+		}
+
+		record->holds[i].notified = true;
+	}
+}
+
+static void
+process_req_new_record(struct lls_config *lls_conf, struct lls_request *req)
+{
+	struct lls_cache *cache = req->cache;
+	char ip_buf[cache->key_str_len];
+	char *ip_str;
+	struct lls_record *record;
+	int ret;
+
+	if (req->ty == LLS_REQ_PUT) {
+		ip_str = cache->ip_str(cache, req->ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, GATEKEEPER,
+			"lls: no record for %s, lcore %u can't put record\n",
+			ip_str == NULL ? cache->name : ip_str,
+			req->u.hold.lcore_id);
+		return;
+	}
+
+	/* This request is starting a new record. */
+	ret = rte_hash_add_key(cache->hash, req->ip_be);
+	if (unlikely(ret == -EINVAL)) {
+		ip_str = cache->ip_str(cache, req->ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, HASH,
+			"Invalid params, could not add record for %s\n",
+			ip_str == NULL ? cache->name : ip_str);
+		return;
+	} else if (unlikely(ret == -ENOSPC)) {
+		ip_str = cache->ip_str(cache, req->ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, HASH, "No space, could not add record for %s\n",
+			ip_str == NULL ? cache->name : ip_str);
+		return;
+	}
+
+	RTE_ASSERT(ret >= 0);
+	record = &cache->records[ret];
+
+	switch (req->ty) {
+	case LLS_REQ_RESOLVE:
+		/* FALLTHROUGH */
+	case LLS_REQ_SUBSCRIBE:
+		/*
+		 * Add all information we have for this new record: that
+		 * it's not valid (it's stale), the IP address,
+		 * an initial timestamp, and the first requester.
+		 */
+		record->map.stale = true;
+		memcpy(record->map.ip_be, req->ip_be, cache->key_len);
+		record->ts = time(NULL);
+		record->holds[0] = req->u.hold;
+		record->num_holds = 1;
+
+		/* Try to resolve record using broadcast. */
+		send_request(lls_conf, cache, req->ip_be, NULL);
+
+		if (lls_conf->debug)
+			lls_cache_dump(cache);
+		return;
+	case LLS_REQ_MODIFY:
+		/* Fill-in new record. */
+		ether_addr_copy(&req->u.mod.ha, &record->map.ha);
+		record->map.port_id = req->u.mod.port_id;
+		record->map.stale = false;
+		memcpy(record->map.ip_be, req->ip_be, cache->key_len);
+		record->ts = req->u.mod.ts;
+		record->num_holds = 0;
+
+		if (lls_conf->debug)
+			lls_cache_dump(cache);
+		return;
+	default:
+		RTE_LOG(ERR, GATEKEEPER,
+			"lls: unrecognized request type (%d)\n", req->ty);
+		return;
+	}
+}
+
+static void
+process_req(struct lls_config *lls_conf, struct lls_request *req)
+{
+	struct lls_record *record;
+	struct lls_cache *cache = req->cache;
+	char ip_buf[cache->key_str_len];
+	char *ip_str;
+
+	int ret = rte_hash_lookup(cache->hash, req->ip_be);
+	if (ret == -ENOENT) {
+		process_req_new_record(lls_conf, req);
+		return;
```

Comment:
It's a matter of style, so feel free to ignore: combine the return and the previous call. The idea is to highlight tail calls.